src/raw/elem.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    convert::{TryFrom, TryInto},
+    time::Duration,
+};
+
+use chrono::{DateTime, TimeZone, Utc};
+
+#[cfg(feature = "decimal128")]
+use super::d128_from_slice;
+use super::{
+    i32_from_slice,
+    i64_from_slice,
+    read_lenencoded,
+    read_nullterminated,
+    u32_from_slice,
+    RawArray,
+    RawDocumentRef,
+    RawError,
+    RawResult,
+};
+use crate::{
+    oid::ObjectId,
+    spec::{BinarySubtype, ElementType},
+    Bson,
+};
+
+/// A BSON value referencing raw bytes stored elsewhere.
+#[derive(Clone, Copy, Debug)]
+pub struct RawBson<'a> {
```

Comment:
Hmm, I have mixed feelings on this. It definitely would be more ergonomic for users, but it semantically doesn't seem quite the same to me. With the existing Document and Bson types, the Document very explicitly contains each Bson value, and with RawDocument and RawBson as written currently in the API, this is also true modulo the ElementType. With the enum as you've written it, the `f64` or `bool` contained in the `RawBson` don't actually exist as part of the RawDocument. I somewhat prefer the idea of having the RawBson values that are accessed through the RawDocument actually being subsets of the RawDocument like with Document and Bson, although I'm having trouble articulating exactly why that is.

Reply:
Indeed, I was actually thinking about an alternative version of RawBson where, from a data standpoint, it's just a slice of a bson document (or array) from the element type byte to the start of the next element type byte.  So a document would be `length + element* + \0`.  Then RawBson could also be made an unsized by-reference type, but then its data would also include the key used to reference it, which is kind of a weird mismatch with how the parsed type works or how people  would expect it to work, based on their experience with JSON.

Reply:
One use case that the current design unlocks that the straight-enum version might break is dealing with not-quite valid BSON data.  I've encountered mongo databases in the wild that contain strings with invalid utf-8 data.  (Specifically it was an ObjectId that somehow got entered as a string.  It was a well-formed string type (had a four-byte length marker), but the data was binary, not utf8.  It was impossible to use python to access that database, because it raised exceptions whenever hitting that data.  Go could handle it because go only pretends that its strings are UTF-8.  The current rust implementation also failed, because it coerced the data to a String immediately on fetching the data.

The current RawBson design decodes the data on request, and provides a way to access the raw bytes as such, specifically to help people work around issues like the one I faced.  It's not particularly ergonomic (kind of intentionally so), but it's *possible*, and that's very valuable.

Reply:
That's a really good point; we actually just added back in support to the library for decoding documents with invalid UTF-8 strings due to a long-standing server bug. I definitely think it's worth giving users a way to deal with data like that in the raw BSON API.

Reply:
> For well-formed BSON, this difference doesn't actually matter in practice, since we return a copy of the various Copy types anyways from the struct, and the lifetime annotation on the enum guarantees that the RawBson instance itself will not outlive the RawDocument it was derived from, even if it doesn't actually contain any reference to it.

I understand that it doesn't matter in terms of safety, but it still feels a little weird. Like I said, I'm having trouble articulating exactly why it bothers me, but it still does.

Reply:
Well anyways, I think this would be a huge usability / familiarity win for users, so unless we have strong reasons to opt for the struct, I would really prefer for us to go the enum route. As I mentioned before, the paths aren't mutually exclusive, so we could do the struct approach too if we really needed to.

Reply:
+1 for the enum route; it's a much more intuitive design IMO, and I think it'll be nice to have some consistency between `Document` and `RawDocument` implementations

Reply:
Per the latest commits, `RawBson` is now an enum. That being said, there is still room for a struct-style `RawBson` via an `Entry`-like API, for example.